src/InArrayValidator.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace ipl\Validator;
+
+use ipl\I18n\Translation;
+
+class InArrayValidator extends BaseValidator
+{
+    use Translation;
+
+    /** @var array The array */
+    protected $haystack;
+
+    /** @var bool Whether the types of the needle in the haystack should also match */
+    protected $strict = false;
+
+    /**
+     * Validates whether the value exists in given haystack
+     *
+     * Optional options:
+     *
+     * - haystack: (array) The array
+     * - strict: (bool) Whether the types of the needle in the haystack should also match, default false
+     *
+     * @param array $options
+     */
+    public function __construct(array $options = [])
+    {
+        if (isset($options['haystack'])) {
+            $this->setHaystack($options['haystack']);
+        }
+
+        $this->setStrict($options['strict'] ?? false);
+    }
+
+    /**
+     * Returns the haystack
+     *
+     * @return array
+     */
+    public function getHaystack(): array
+    {
+        return $this->haystack ?? [];
+    }
+
+    /**
+     * Set the haystack
+     *
+     * @param  array $haystack
+     *
+     * @return self
+     */
+    public function setHaystack(array $haystack): self
+    {
+        $this->haystack = $haystack;
+
+        return $this;
+    }
+
+    /**
+     * Whether the types of the needle in the haystack should also match
+     *
+     * @return bool
+     */
+    public function isStrict(): bool
+    {
+        return $this->strict;
+    }
+
+    /**
+     * Set whether the types of the needle in the haystack should also match
+     *
+     * @param  bool $strict
+     *
+     * @return self
+     */
+    public function setStrict(bool $strict = true): self
+    {
+        $this->strict = $strict;
+
+        return $this;
+    }
+
+    public function isValid($value)
+    {
+        // Multiple isValid() calls must not stack validation messages
+        $this->clearMessages();
+
+        $haystack = $this->getHaystack();
```

Comment:
What about something like this?

```
// Class method:
$looseEqual = function ($a, $b): bool {
    return $a == $b;
};

// Class method:
$strictEqual = function ($a, $b): bool {
    return $a === $b;
};

$notInArray = array_udiff((array) $value, $haystack, $strict ? $strictEqual : $looseEqual);
```

Supports strict, non-arrays and arrays as needle and is more readable.

Reply:
There's the spaceship operator to achieve this: `<=>`
There's no strict version though.

Reply:
You are absolutely right about the `int`. But such callbacks usually return `-1`, `0` or `1`. Therefore I don't understand the `99`. Please also make it possible to test the diff.

Your current code will yield wrong results:

```
$validator = (new InArrayValidator())
    ->setHaystack(['a', 'b', 'c']);

var_dump($validator->isValid(['a', 'd']), $validator->getMessages());
```

The comparison functions can be static methods.

Reply:
Are the diff array keys also of interest? Can't find a use case for it. I assume we only want the values.

Reply:
Only values.

Reply:
the array_udiff comparison function returns unexpected results, especially with strict check. Discussed with @nilmerg and we decided to add the old variant again.